text/0008-kadalu-kubernetes-storage-monitoring.adoc

```diff
@@ -0,0 +1,153 @@
+---
+start_date: 2021-08-29
+rfc_pr: https://github.com/kadalu/rfcs/pull/0000
+status: SUBMITTED
+available_since: (leave this empty)
+---
+
+= Monitoring Kadalu Kubernetes Storage
+
+== Summary
+
+Kadalu Storage pods are distributed across the nodes and it is very difficult to collect the metrics from all the Pods and aggregate the details to makes sense of the data. Even though each Pods are capable of exporting the metrics and set the necessary annotations required to make it discover by Prometheus server, it is very difficult to add Prometheus rules and alerts without having the full knowledge of the Gluster Cluster. It is also difficult to fetch these metrics if a non-Prometheus application is interested monitor the Cluster.
+
+The proposed solution collects the metrics from all the Pods managed by Kadalu Operator and exports the *Metrics* and *Events* from single API.
+
+== Authors
+
+- Aravinda Vishwanathapura <aravinda@kadalu.io>
+
+
+== Motivation
+
+- Business logic in one place - If the metrics are exported by individual pods then business logic should be added in the Prometheus server to make decisions.
+- Get all metrics in one place.
+- Events and Alerts - Kadalu Operator can generate events and alerts based on the metrics collected.
+- Kadalu Operator gets more power - Kadalu Operator can take better decisions for the internal maintenance activities depending on the Cluster state. For example, smart rolling upgrade based on Heal pending and Server pods health metrics.
+- Supports non-Prometheus consumers.
+- Additional information export for non-Prometheus API users - For example, list of Volume options configured, Version information etc.
+- External monitoring integration is easy - Since all metrics are available in single place, it will be easy to integrate the metrics with external Cloud hosted monitoring solutions.
+
+== Detailed design
+
+Kadalu Operator runs `kubectl get pods -n kadalu` to get the list of all the resources available in Kadalu namespace. Additionally it fetches the nodes list and all the Storage information from the ConfigMap. With these information a few metrics will be derived as follows.
+
+- Number of Up CSI node plugins by comparing the list of nodes and the list returned by `get pods` command.
```

Comment:
- Might need to adjust the metrics or a note corresponding to taints & tolerations on nodes

Reply:
Ack. To start with we can show, `up_node_plugins` or something.